### PR TITLE
feat: Enable remote invoke support for Step Functions, SQS, and Kinesis data streams

### PR DESCRIPTION
**Pull request by hnnasit**
#### Description

This PR enables `sam remote invoke` support for SFN, SQS, and Kinesis data streams. The following changes are being made:
- Enable test-event for only Lambda functions
- Update `--parameter` help text to include new services
- Update `examples` in the help text
- Define a priority order for default resources used with remote invoke. E.g. If 1 Lambda function is present, remote invoke will pick it up from the stack without providing resource_id
- Update sam-cli.json schema file
- Enable executors for new services and in remote_invoke_context
- Enable all remote invoke integration tests

The examples section of the CLI help text:
```
Examples:

    Lambda Functions:

        Invoke default lambda function with empty event:
            $samdev remote invoke --stack-name hello-world
                                            

        Invoke default lambda function with event passed as text input:
            $samdev remote invoke --stack-name hello-world -e '{"message": "hello!"}'
                                            

        Invoke named lambda function with an event file:
            $samdev remote invoke --stack-name hello-world HelloWorldFunction --event-file event.json
                                            

        Invoke function with event as stdin input:
            $ echo '{"message": "hello!"}' | samdev remote invoke HelloWorldFunction --event-file -
                                            

        Invoke function using lambda ARN and get the full AWS API response:
            $samdev remote invoke arn:aws:lambda:us-west-2:123456789012:function:my-function -e <> --output json
                                            

        Asynchronously invoke function with additional boto parameters:
            $samdev remote invoke HelloWorldFunction -e <> --parameter InvocationType=Event --parameter Qualifier=MyQualifier
                                            

        Dry invoke a function to validate parameter values and user/role permissions:
            $samdev remote invoke HelloWorldFunction -e <> --output json --parameter InvocationType=DryRun
                                            

    Step Functions:

        Start execution with event passed as text input:
            $samdev remote invoke --stack-name mock-stack StockTradingStateMachine -e '{"message": "hello!"}'
                                            

        Start execution using its physical-id or ARN with an execution name parameter:
            $samdev remote invoke arn:aws:states:us-east-1:123456789012:stateMachine:MySFN -e <> --parameter name=mock-execution-name
                                            

        Start execution with an event file and get the full AWS API response:
            $samdev remote invoke --stack-name mock-stack StockTradingStateMachine --event-file event.json --output json
                                            

        Start execution with event as stdin input and pass the X-ray trace header to the execution:
            $ echo '{"message": "hello!"}' | $samdev remote invoke --stack-name mock-stack StockTradingStateMachine --parameter traceHeader=<>
                                            

    SQS Queue:

        Send a message with the MessageBody passed as event:
            $samdev remote invoke --stack-name mock-stack MySQSQueue -e hello-world
                                            

        Send a message using its physical-id and pass event using --event-file:
            $samdev remote invoke https://sqs.us-east-1.amazonaws.com/12345678910/QueueName --event-file event.json
                                            

        Send a message using its ARN and delay the specified message:
            $samdev remote invoke arn:aws:sqs:region:account_id:queue_name -e hello-world --parameter DelaySeconds=10
                                            

        Send a message along with message attributes and get the full AWS API response:
            $samdev remote invoke --stack-name mock-stack MySQSQueue -e hello-world --output json --parameter MessageAttributes='{"City": {"DataType": "String", "StringValue": "City"}}'
                                            

    Kinesis Data Stream:

        Put a record using the data provided as event:
            $samdev remote invoke --stack-name mock-stack MyKinesisStream -e '{"message": "hello!"}'
                                            

        Put a record using its physical-id and pass event using --event-file:
            $samdev remote invoke MyKinesisStreamName --event-file event.json
                                            

        Put a record using its ARN and override the key hash:
            $samdev remote invoke arn:aws:kinesis:us-east-2:123456789012:stream/mystream --event-file event.json --parameter ExplicitHashKey=<>
                                            

        Put a record with a sequence number for ordering with a PartitionKey:
            $samdev remote invoke MyKinesisStreamName --event hello-world --parameter SequenceNumberForOrdering=<> --parameter PartitionKey=<>
```

`--parameter` help text:
```
Additional parameters that can be passed to invoke the resource.
                                    
Lambda Function(Buffered stream): The following additional parameters can be used to
invoke a lambda resource and get a buffered response:
InvocationType='Event'|'RequestResponse'|'DryRun', LogType='None'|'Tail',
ClientContext='base64-encoded string' Qualifier='string'.

Lambda Function(Response stream): The following additional parameters can be used to
invoke a lambda resource with response streaming:
InvocationType='RequestResponse'|'DryRun', LogType='None'|'Tail',
ClientContext='base64-encoded string', Qualifier='string'.

Step Functions: The following additional parameters can be used to start a state
machine execution: name='string', traceHeader='string'

SQS Queue: The following additional parameters can be used to send a message to an
SQS queue: DelaySeconds=integer, MessageAttributes='json string',
MessageSystemAttributes='json string', MessageDeduplicationId='string',
MessageGroupId='string'

Kinesis Data Stream: The following additional parameters can be used to put a record
in the kinesis data stream: PartitionKey='string', ExplicitHashKey='string',
SequenceNumberForOrdering='string', StreamARN='string'                                 
```

#### Why is this change necessary?
To enable new services support for remote invoke

#### How does it address the issue?
Described above

#### What side effects does this change have?
No side effects

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [x] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/commands/remote/remote_invoke_context.py [hunk 1]
```diff
@@ -164,14 +175,14 @@ def _get_single_resource_from_stack(self) -> CloudFormationResourceSummary:
         """
         LOG.debug(
             "Trying to get single resource with %s type in % stack since no resource id is provided",
```
Comment (lucashuy) [inline samcli/commands/remote/remote_invoke_context.py:177]: Missing an `s` for the second replacement string in the debug?

**Review by lucashuy — APPROVED** [review]

File: samcli/commands/_utils/options.py [hunk 2]
```diff
@@ -587,13 +587,22 @@ def remote_invoke_parameter_click_option():
...
-        help="Additional parameters that can be passed to invoke the resource.\n"
-        "The following additional parameters can be used to invoke a lambda resource and get a buffered response: "
-        "InvocationType='Event'|'RequestResponse'|'DryRun', LogType='None'|'Tail', "
-        "ClientContext='base64-encoded string' Qualifier='string'. "
-        "The following additional parameters can be used to invoke a lambda resource with response streaming: "
-        "InvocationType='RequestResponse'|'DryRun', LogType='None'|'Tail', "
-        "ClientContext='base64-encoded string', Qualifier='string'.",
+        help="Additional parameters that can be passed"
+        " to invoke the resource.\n\n"
+        "Lambda Function(Buffered stream): The following additional parameters can be used to invoke a lambda resource"
```
Comment (mndeveci) [inline samcli/commands/_utils/options.py:592]:
```suggestion
        "Lambda Function (Buffered stream): The following additional parameters can be used to invoke a lambda resource"
```

File: samcli/commands/_utils/options.py [hunk 3]
```diff
@@ -587,13 +587,22 @@ def remote_invoke_parameter_click_option():
...
+        " and get a buffered response: InvocationType='Event'|'RequestResponse'|'DryRun', LogType='None'|'Tail', "
+        "ClientContext='base64-encoded string' Qualifier='string'.\n\n"
+        "Lambda Function(Response stream): The following additional parameters can be used to invoke a lambda resource"
```
Comment (mndeveci) [inline samcli/commands/_utils/options.py:595]:
```suggestion
        "Lambda Function (Response stream): The following additional parameters can be used to invoke a lambda resource"
```

File: samcli/commands/remote/remote_invoke_context.py [hunk 4]
```diff
@@ -180,10 +191,21 @@ def _get_single_resource_from_stack(self) -> CloudFormationResourceSummary:
...
+                    if resource_summary.resource_type == resource_type:
+                        resource_type_count += 1
+                        single_resource_summary = resource_summary
+                    if resource_type_count > 1:
+                        raise AmbiguousResourceForRemoteInvoke(
+                            f"{self._stack_name} contains more than one resource that could be used with remote invoke,"
+                            f" please provide resource_id argument to resolve ambiguity."
+                        )
+                if resource_type_count == 1 and single_resource_summary:
+                    return single_resource_summary
```
Comment (mndeveci) [inline samcli/commands/remote/remote_invoke_context.py:208]: nit: can we eliminate some of the indentation here? 🤔

**Review by mndeveci — APPROVED** [review]
LGTM! 🥳 Left some small comments here and there

**Pull request merged**